### PR TITLE
Fix bit overflow with softmax

### PR DESCRIPTION
**Pull request by calad0i**
A# Description

> :memo: Please include a summary of the change.
> 
> * Please also include relevant motivation and context.
> * List any dependencies that are required for this change.

Fix for #885 : softmax working with less than 10 bits of input/table size. Tested with Vivado/Vitis/quartus.

## Type of change

For a new feature or function, please create an issue first to discuss it
with us before submitting a pull request.

Note: Please delete options that are not relevant.

- [x] Bug fix (non-breaking change that fixes an issue)

## Tests

> :memo: Please describe the tests that you ran to verify your changes.
>
> * Provide instructions so we can reproduce.
> * Please also list any relevant details for your test configuration.

**Test Configuration**:

`pytest/test_softmax.py`

## Checklist

- [x] I have read the [guidelines for contributing](https://github.com/fastmachinelearning/hls4ml/blob/main/CONTRIBUTING.md).
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings. - Raise a warning when the added fix is triggered.
- [x] I have installed and run `pre-commit` on the files I edited or added.
- [x] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

**Comment by jmitrevs** [comment]
I wonder if the optimizer should really be in the backends directory. I think optimizers in `model/optimizers` should not concern themselves with backends.

**Review by jmitrevs — CHANGES_REQUESTED** [review]
I wonder if the optimizer should really be in the backends directory. I think optimizers in `model/optimizers` should not concern themselves with backends.

**Comment by calad0i** [comment]
> I wonder if the optimizer should really be in the backends directory. I think optimizers in `model/optimizers` should not concern themselves with backends.

Moved

File: test/pytest/test_softmax.py [hunk 1]
```diff
@@ -10,50 +10,47 @@
...
-
-
-def high_accuracy_distribution(shape):
-    '''Start with a flat distribution, then pick a random member of each row to amplify'''
-    x = np.random.rand(*shape)
-    imax = np.random.randint(0, shape[1], size=shape[0])
-    x[:, imax] *= 10
-    return x
+def normal_dist(shape):
+    return np.clip(np.random.normal(0, 8, shape), -32, 31)
```
Comment (vloncar) [inline test/pytest/test_softmax.py:14]: Why are we getting rid of existing distributions?

File: test/pytest/test_softmax.py [hunk 2]
```diff
@@ -10,50 +10,47 @@
...
-def test_softmax(backend, strategy, generate_data, input_shape, io_type, function):
+def test_softmax(backend, strategy, generate_data, input_bits, input_shape, table_bits, io_type):
     X = generate_data
     model = tf.keras.models.Sequential()
     model.add(tf.keras.layers.Activation(input_shape=input_shape, activation='softmax', name='softmax'))
     model.compile()
 
-    f_type = 'ac_fixed<18,8,true,AC_RND,AC_SAT>' if backend == 'Quartus' else 'ap_fixed<18,8,AP_RND,AP_SAT>'
+    f_type = (
+        f'ac_fixed<{table_bits},true,AC_RND,AC_SAT>' if backend == 'Quartus' else f'ap_fixed<{table_bits},AP_RND,AP_SAT>'
```
Comment (vloncar) [inline test/pytest/test_softmax.py:43]: I think a new standard for tests that use multiple backends is to just do `fixed<W,I>`

File: test/pytest/test_softmax.py [hunk 3]
```diff
@@ -10,50 +10,47 @@
...
+    )
     cfg = hls4ml.utils.config_from_keras_model(model, granularity='name')
     cfg['LayerName']['softmax']['Strategy'] = strategy
     cfg['LayerName']['softmax']['inv_table_t'] = f_type
     cfg['LayerName']['softmax']['exp_table_t'] = f_type
+    cfg['LayerName']['softmax_input']['Precision']['result'] = f'ap_fixed<{input_bits}>'
```
Comment (vloncar) [inline test/pytest/test_softmax.py:49]: Related to the comment above, here you're actually relying on what is essentially a bug to correctly parse even in case of Quartus.

File: test/pytest/test_softmax.py [hunk 4]
```diff
@@ -92,7 +89,6 @@ def test_softmax_skipped(backend, io_type):
     assert len(hls_layers) == 2
 
     # Verify hls4ml output is equal to Dense output
-    y_keras = model.predict(X)
-    y_hls4ml = hls_model.predict(X).reshape(y_keras.shape)
-    keras_trace = hls4ml.model.profiling.get_ymodel_keras(model, X)
-    np.testing.assert_allclose(y_hls4ml, keras_trace['dense'], rtol=0, atol=2e-2)
+    y_keras = model.layers[0](X).numpy()  # type: ignore
```
Comment (vloncar) [inline test/pytest/test_softmax.py:92]: Why do you need to do it like this?

File: test/pytest/test_softmax.py (hunk 1 above)
Comment (calad0i) [reply test/pytest/test_softmax.py:14]: The existing distribution generation does not feel necessary, and I don't see a good reason to run tests twice on uniform and uniform+manual outliers. A gaussian will also give variables of many scales, giving some outliers by default.

File: test/pytest/test_softmax.py (hunk 2 above)
Comment (calad0i) [reply test/pytest/test_softmax.py:43]: These are inherited from the old test file. Will `SAT` / `RND` flags be also auto-converted? Will change.

File: test/pytest/test_softmax.py (hunk 4 above)
Comment (calad0i) [reply test/pytest/test_softmax.py:92]: The original code was redundant (and confusing to me): what is done here is essentially skipping the last layer in Keras. As there are only two, applying the first layer only is enough.
Will revert if you think the original one is better.

Comment (vloncar) [reply test/pytest/test_softmax.py:92]: I think we want just `model.predict(X)`. The section looks like it was cut from a different test. We should use the opportunity to clean up (also the surrounding comments)

File: test/pytest/test_softmax.py (hunk 1 above)
Comment (calad0i) [reply test/pytest/test_softmax.py:14]: In this case I shall purge this part, should an equivalent test already exists somewhere.

File: test/pytest/test_softmax.py (hunk 4 above)
Comment (calad0i) [reply test/pytest/test_softmax.py:92]: In this case I shall purge this part, should an equivalent test already exists somewhere.

**Review by jmitrevs — APPROVED** [review]
